File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        
        

        if not root:
            return False 

        bfs , s = [root] , set()

        for node  in bfs:
            if k - node.val in s:
                return True 
            s.add(node.val) 
            if node.left:
                bfs.append(node.left)
            if node.right:
                bfs.append(node.right)
        return False 

File: 653-two-sum-iv-input-is-a-bst/653-two-sum-iv-input-is-a-bst.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

void traverse(struct TreeNode* root, int *arr, int *idx){ 

    if(!root) return ; 
    traverse( root->left, arr, idx) ; 
    arr[(*idx)++ ] = root->val; 
    traverse( root->right , arr , idx) ; 
}


bool findTarget(struct TreeNode* root, int k){

    
        
    int *arr = malloc(sizeof(int)*10000) , idx = 0 , sumv=0 ; 
    traverse(root, arr , &idx) ; 
    int i,j ; 
    for(i=0 ,j=idx-1; i<j; ){

        sumv = arr[i]+arr[j] ; 
        if(sumv > k) j  -- ; 
        else if(sumv < k) i++ ; 
        else break ; 
    }
    return i==j? false:true ; 

}